### PR TITLE
Create PublicKey wrapper for AttestationAuthoritySpec

### PR DESCRIPTION
**Pull request by acamadeo**
Create `PublicKey`, a wrapper that stores public key data for PGP or PKIX public keys. Update `AttestationAuthoritySpec` to include a list of `PublicKeys`. Modify Attestation verification algorithm to validate and use `PublicKeys`. Update documentation with instructions on creating the new `AttestationAuthoritySpec` using `PublicKeys`. Update scripts and testfiles to use new `AttestationAuthoritySpec`.

### CODE REVIEW

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 1]
```diff
@@ -33,9 +39,31 @@ type AttestationAuthority struct {
...
+// PublicKey stores key data used to verify Attestations.
+type PublicKey struct {
+	// KeyId is the ID of this public key. This is a required field for all keys.
+	KeyId string `json:"keyId"`
+	// KeyType is the type of this public key. It should be one of "PGP_KEY" or
+	// "PKIX_KEY".
+	KeyType string `json:"keyType`
+	// PgpPublicKey is the base64-encoded payload for a PGP public key. Provide
+	// this directly as a string.
+	PgpPublicKey string `json:"pgpPublicKey,omitemtpy"`
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Done.

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 2]
```diff
@@ -33,9 +39,31 @@ type AttestationAuthority struct {
...
-	PolicyType    string   `json:"policyType"`
+	NoteReference string      `json:"noteReference"`
+	PublicKeys    []PublicKey `json:"publicKeys"`
+	PolicyType    string      `json:"policyType"`
+}
+
+// PublicKey stores key data used to verify Attestations.
+type PublicKey struct {
+	// KeyId is the ID of this public key. This is a required field for all keys.
+	KeyId string `json:"keyId"`
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Apologies, I was basing the draft PR off of the wrong notes. We decided that this would be optional for PGP keys. I have made the necessary changes to implement this

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 3]
```diff
@@ -33,9 +39,31 @@ type AttestationAuthority struct {
...
+	// PkixPublicKey stores data for a PKIX public key.
+	PkixPublicKey PkixPublicKey `json:"pkixPublicKey,omitempty"`
+}
+
+type PkixPublicKey struct {
+	// PublicKey is the payload for a PKIX public key. Provide this as a string.
+	PublicKey string `json:"publicKey"`
+	// SignatureAlgorithm is the type of algorithm that was used to generate
+	// the signature.
+	SignatureAlgorithm string `json:"signatureAlgorithm"`
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Done, I left this as a TODO for the next PR.

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 4]
```diff
@@ -33,9 +39,31 @@ type AttestationAuthority struct {
...
+	// KeyId is the ID of this public key. This is a required field for all keys.
+	KeyId string `json:"keyId"`
+	// KeyType is the type of this public key. It should be one of "PGP_KEY" or
+	// "PKIX_KEY".
+	KeyType string `json:"keyType`
+	// PgpPublicKey is the base64-encoded payload for a PGP public key. Provide
+	// this directly as a string.
+	PgpPublicKey string `json:"pgpPublicKey,omitemtpy"`
+	// PkixPublicKey stores data for a PKIX public key.
+	PkixPublicKey PkixPublicKey `json:"pkixPublicKey,omitempty"`
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Done.

File: pkg/kritis/review/validating_transport.go [hunk 5]
```diff
@@ -40,18 +41,69 @@ type AttestorValidatingTransport struct {
 	Attestor v1beta1.AttestationAuthority
 }
 
-func (avt *AttestorValidatingTransport) GetValidatedAttestations(image string) ([]attestation.ValidatedAttestation, error) {
-	keys := map[string]string{}
-	for _, keyData := range avt.Attestor.Spec.PublicKeyList {
-		key, fingerprint, err := secrets.KeyAndFingerprint(keyData)
+func (avt *AttestorValidatingTransport) ValidatePublicKey(pubKey v1beta1.PublicKey) error {
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: It is not used outside. I changed it to unexported.

File: pkg/kritis/review/validating_transport_test.go [hunk 6]
```diff
@@ -38,8 +39,8 @@ func encodeB64(in string) string {
 func TestValidatingTransport(t *testing.T) {
```
Comment (ooq) [inline pkg/kritis/review/validating_transport_test.go]: I'm a bit concerned that this test suite has grown to such a size that it's hard to follow now. Maybe we should add type-specific testing elsewhere so this suite can focus more on the main logic?

File: integration/testdata/generic-attestation-policy/test-attestor-1.yaml [hunk 7]
```diff
@@ -4,6 +4,8 @@ metadata:
...
+  publicKeys:
+    keyType: PGP_KEY
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply integration/testdata/generic-attestation-policy/test-attestor-1.yaml]: Done.

File: integration/testdata/generic-attestation-policy/test-attestor-1.yaml [hunk 8]
```diff
@@ -4,6 +4,8 @@ metadata:
...
...CcE9IYmFndXlpMTFUV2RmTXo4LzNBSmtJUUo3NVFyZklEUzljNE9ZUWJ3OVlRSGEzUFdrai9q UApFM3FRY3lGMnQ2Y05ET0NNcmhCMEtvYjBvWWwrcGVoamFhMHRtM3FiaTdjU1p5TEkxL2dxMjJv MHZGVVM3eUZzCnBhNzhjYi9TSUhGUFJSZGFVVHJpdXIvclhhMm1IaE9qQU1zT09jVmFldkt0WkNM bFBCZThmVXZRZ1FJVnpRaEwKZHVRcjBtemlOYk56QXZoS0ZKTXRTdU5tUFFxSlJyWGFNNldMNjk3 eWJYNjdaOCtUQkI5NXd3QVJBUUFCaVFHMgpCQmdCQ2dBZ0ZpRUVzWlM0eXVXVk9iUjBlYlF1VUR0 aXAxR0NjaUVGQWw1eVVQRUNHd3dBQ2drUVVEdGlwMUdDCmNpSEphQXdBb2dTeDdlcFM1U2o1Ymh1 eDIvdFhLYncxa2dwbkZVeWRuSm5SMjVscE50TVBRbFNMMjFOZCtZZFcKU3h1c0dTNUV2WDQvald6 eVk1ckNyV3hiOWNNcFV4MS8wUjg0Zi9HNGxULzRCR2o4NEtud0JXWmlaTzNIbXhCVQpoaEZzRVZv dk9sVjg2VlNrQVBOYVp3REJBVlhGV3Q0MjBJNm8vM0hVSkdyOHpqOFl4ZEV1S09zb2wzOTA5VjR5 ClM0VEltTW5VQWtaOTZCMks2OFRmSEdsR0p1dEhiZHYwMU5Yc0JKOXI2OVJMa201SmNQS0xkdEYv YklKVjVNTjEKQzV3VFNTMTVsaldGMHJKZ1FVclVyOGQydGMyZWN5Yks2djRKQUo3cncvOE5Od05j eEJGRlVMQ0hZZlZoMFBnRQp1Q0R3cEdTM21jZjdWVEwvMTBKbEVTbFR6NkdxWVpzMnlGWGJsQnZT MVRQSDA3VE96VEg1YlN1OHh5eVNuR0d3CjcxWHk3c0crSng2VURZWFlOazBtL2p2TjJSeTBZemxN Y3NCT284dHAxNDUxVVgwRnoybzlSRFZLejFCQlpFM0gKRTIwT3lKMk5HWEhyUlRYZzZnNjRVbDZh ZUI1aHlZUk9Yc1A1ZEtFVDVNK1hiUWhRbUJNaVZ3bitNTlZWbW8wagp1bjhiMDNBMwo9eHBkUAot LS0tLUVORCBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCg==
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply integration/testdata/generic-attestation-policy/test-attestor-1.yaml]: Done. Had to remove the newlines.

File: integration/testdata/generic-attestation-policy/test-attestor-2.yaml [hunk 9]
```diff
@@ -4,6 +4,8 @@ metadata:
...
+  publicKeys:
+    keyType: PGP_KEY
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply integration/testdata/generic-attestation-policy/test-attestor-2.yaml]: Done.

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 10]
```diff
@@ -20,6 +20,12 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	// Key Types
+	PgpKeyType  = "PGP_KEY"
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Done. Removed "_KEY"

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 11]
```diff
@@ -33,9 +39,37 @@ type AttestationAuthority struct {
...
-	PublicKeyList []string `json:"publicKeyList"`
-	PolicyType    string   `json:"policyType"`
+	NoteReference string      `json:"noteReference"`
+	PublicKeys    []PublicKey `json:"publicKeys"`
+	PolicyType    string      `json:"policyType"`
+}
+
+// PublicKey stores key data used to verify Attestations.
+type PublicKey struct {
+	// KeyId is the ID of this public key (required for PKIX keys). For PKIX
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: After our discussion with Nenad, we decided that the `keyId` field would be required for PKIX keys. We are unable to achieve full compatibility with Binauthz in this regard, because Kritis lacks a policy management service (PMS). 

Binauthz's PMS can create its own PKIX `keyId` that it can return to the users. Since Kritis has no such system, we must ask that our users define and refer to their keys through an ID that they provide.

File: pkg/kritis/review/validating_transport.go [hunk 12]
```diff
@@ -40,20 +41,71 @@ type AttestorValidatingTransport struct {
...
+	if err := avt.validatePublicKeyId(pubKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validatePublicKeyType(pubKey v1beta1.PublicKey) error {
+	switch pubKey.KeyType {
+	case v1beta1.PgpKeyType:
+		if pubKey.PkixPublicKey != (v1beta1.PkixPublicKey{}) {
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: According to [this](https://golang.org/pkg/encoding/json/#Marshal), `omitempty` ensures that fields with empty values are omitted when encoded into JSON. But in our case, we are parsing user-provided JSON (technically YAML) into our structs. `omitempty` does not provide any guarantees in this situation, so we have to check manually

File: pkg/kritis/review/validating_transport.go [hunk 13]
```diff
@@ -40,20 +41,71 @@ type AttestorValidatingTransport struct {
...
+			return fmt.Errorf("Invalid PGP key: %v. PkixPublicKey field should not be set", pubKey)
+		}
+	case v1beta1.PkixKeyType:
+		if pubKey.AsciiArmoredPgpPublicKey != "" {
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: See response above.

File: pkg/kritis/review/validating_transport.go [hunk 14]
```diff
@@ -40,20 +41,71 @@ type AttestorValidatingTransport struct {
...
+		return err
+	}
+	return nil
+}
+
+func validatePublicKeyType(pubKey v1beta1.PublicKey) error {
+	switch pubKey.KeyType {
+	case v1beta1.PgpKeyType:
+		if pubKey.PkixPublicKey != (v1beta1.PkixPublicKey{}) {
+			return fmt.Errorf("Invalid PGP key: %v. PkixPublicKey field should not be set", pubKey)
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: So in Binauthz, users are expected not to provide a `keyId` for PGP keys -- the PMS computes one based on the key's OpenPGP fingerprint and attaches it to the resource. Without a PMS, we cannot enforce this behavior for Kritis users. But we want to encourage them to do things the same way as in Binauthz.

Consequently, we are encouraging users to either provide the OpenPGP fingerprint as the `keyId` or leave it blank, as they would in a Binauthz policy. Thus, `pubKey.AsciiArmoredPgpPublicKey` is allowed to be empty.

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go [hunk 15]
```diff
@@ -33,9 +39,37 @@ type AttestationAuthority struct {
...
+	// keys, please use a RFC3986 conformant URI. For PGP keys, this field
+	// should be the OpenPGP RFC4880 V4 fingerprint of the key or be left
+	// blank.
```
Comment (nenaddedic) [inline pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Please document how blank is interpreted.

File: docs/resources.md [hunk 16]
```diff
@@ -229,17 +229,37 @@ metadata:
...
     - ...
     - ...
 ```
 
-Where “image-attestor” is the project for creating AttestationAuthority Notes.
+where “image-attestor” is the project for creating AttestationAuthority Notes.
 
 In order to create notes, the service account `gac-ca-admin` must have `containeranalysis.notes.attacher role` on this project.
 
 The Kubernetes secret `foo` must have data fields `private` and `public` which contain the gpg private and public key respectively.
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done.

File: docs/resources.md [hunk 17]
```diff
@@ -229,17 +229,37 @@ metadata:
...
 
-`publicKeyList` is a list of base64 encoded PEM public keys for the gpg secret.
-The list is used to support key rotations. 
+
+`publicKeys` is a list of public keys for the Kubernetes secret. Key rotation is supported by listing multiple keys.
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done.

File: docs/resources.md [hunk 18]
```diff
@@ -229,17 +229,37 @@ metadata:
...
 An image is attested if it has an attestation verifiable by ANY of the public keys.
+
+
+Two types of public keys are supported: PGP keys and PKIX keys.
+All public keys should specify their `keyType` (either `PGP_KEY` or `PKIX_KEY`).
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done.

File: docs/resources.md [hunk 19]
```diff
@@ -229,17 +229,37 @@ metadata:
...
 The Kubernetes secret `foo` must have data fields `private` and `public` which contain the gpg private and public key respectively.
 
-`publicKeyList` is a list of base64 encoded PEM public keys for the gpg secret.
-The list is used to support key rotations. 
+
+`publicKeys` is a list of public keys for the Kubernetes secret. Key rotation is supported by listing multiple keys.
 An image is attested if it has an attestation verifiable by ANY of the public keys.
+
+
+Two types of public keys are supported: PGP keys and PKIX keys.
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done.

File: docs/resources.md [hunk 20]
```diff
@@ -229,17 +229,37 @@ metadata:
...
+All public keys should specify their `keyType` (either `PGP_KEY` or `PKIX_KEY`). 
+PKIX keys MUST specify a valid RFC3986 `keyId`. For PGP keys, `keyId` should either be the OpenPGP RFC4880 V4 fingerprint of the 
+key payload or be left blank. 
+
+For PGP keys, please provide the base64-encoded PEM public key under `asciiArmoredPgpPublicKey`.
+PKIX key data should be provided under `pkixPublicKey`. It should include the `signatureAlgorithm` used to generate the key,
+and the `publicKey` payload.
+
+Note that keys should contain either `asciiArmoredPgpPublicKey` or `pkixPublicKey`. Keys that specify both fields are disregarded.
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done.

File: integration/run_test.go [hunk 21]
```diff
@@ -291,8 +292,13 @@ func setUpISP(t *testing.T) (kubernetes.Interface, *v1.Namespace, func(t *testin
 	pubKey, privKey := testutil.CreateKeyPair(t, aaSecret)
 	// Create key secret in k8s cluster
 	createKeySecret(t, aaSecret, ns.Name, pubKey, privKey)
+
+	pgpKey, err := secrets.NewPgpKey(privKey, "", pubKey)
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply integration/run_test.go]: I think there could be some problems with that approach. We may end up extending this function to create a PKIX key pair. Though we could deterministically compute an `id`, it may not match the `id` provided by the user, and thus it would go unused.

File: pkg/kritis/review/validating_transport.go [hunk 22]
```diff
@@ -40,20 +41,71 @@ type AttestorValidatingTransport struct {
 	Attestor v1beta1.AttestationAuthority
 }
 
-func (avt *AttestorValidatingTransport) GetValidatedAttestations(image string) ([]attestation.ValidatedAttestation, error) {
-	keys := map[string]string{}
-	numKeys := len(avt.Attestor.Spec.PublicKeyList)
-	for i, keyData := range avt.Attestor.Spec.PublicKeyList {
-		key, fingerprint, err := secrets.KeyAndFingerprint(keyData)
+func (avt *AttestorValidatingTransport) validatePublicKey(pubKey v1beta1.PublicKey) error {
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: Done in a follow-up PR.

File: pkg/kritis/review/validating_transport_test.go (hunk 6 above)
Comment (acamadeo) [reply pkg/kritis/review/validating_transport_test.go]: Done in a follow-up PR.

File: docs/resources.md [hunk 23]
```diff
@@ -229,17 +229,37 @@ metadata:
     namespace: qa
 spec:
     noteReference: projects/image-attestor/notes/qa-note
-    publicKeyList:
+    publicKeys:
+    - keyType: PGP_KEY
+      keyId: ...
+      asciiArmoredPgpPublicKey: ...
+    - keyType: PKIX_KEY
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done. Added a comment to this yaml example, and added a bolded note under a new "PKIX Public Keys" subsection

File: pkg/kritis/review/validating_transport.go [hunk 24]
```diff
@@ -40,20 +41,71 @@ type AttestorValidatingTransport struct {
...
+	if err := validatePublicKeyType(pubKey); err != nil {
+		return err
+	}
+	if err := avt.validatePublicKeyId(pubKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validatePublicKeyType(pubKey v1beta1.PublicKey) error {
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: Done.

File: pkg/kritis/review/validating_transport.go [hunk 25]
```diff
@@ -40,20 +41,71 @@ type AttestorValidatingTransport struct {
...
+		if pubKey.PkixPublicKey == (v1beta1.PkixPublicKey{}) {
+			return fmt.Errorf("Invalid PKIX key: %v. PkixPublicKey field should be set", pubKey)
+		}
+	default:
+		return fmt.Errorf("Unsupported key type %s for key %v", pubKey.KeyType, pubKey)
+	}
+	return nil
+}
+
+func (avt *AttestorValidatingTransport) validatePublicKeyId(pubKey v1beta1.PublicKey) error {
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: Done.

File: pkg/kritis/apis/kritis/v1beta1/attestationauthority.go (hunk 15 above)
Comment (acamadeo) [reply pkg/kritis/apis/kritis/v1beta1/attestationauthority.go]: Done.

File: docs/resources.md [hunk 26]
```diff
@@ -246,20 +249,36 @@ where “image-attestor” is the project for creating AttestationAuthority Note
...
+### PGP Public Keys
+
+| Field    | Type    | Value   |
+|----------|----------|---------------|
+| `keyType`  | string   | "PGP"| 
+| `keyId`  | string   | Optional: either OpenPGP RFC4880 V4 fingerprint of the key payload or blank. If left blank, the `keyId` will be computed as the key's OpenPGP fingerprint.| 
+| `asciiArmoredPgpPublicKey`  | string | A base64-encoded ASCII-armored representation of a PGP public key, as the entire output by the command `gpg --export --armor foo@example.com | base64` (either LF or CRLF line endings). |
+| `pkixPublicKey` | PkixPublicKey object | empty|
 
+### PKIX Public Keys
```
Comment (nenaddedic) [inline docs/resources.md]: Optional: Perhaps it's enough to only say that this is not implemented yet.  Your call...

Comment (acamadeo) [reply docs/resources.md]: Done.

File: docs/resources.md [hunk 27]
```diff
@@ -246,20 +249,36 @@ where “image-attestor” is the project for creating AttestationAuthority Note
...
 In order to create notes, the service account `gac-ca-admin` must have `containeranalysis.notes.attacher role` on this project.
 
-The Kubernetes secret `foo` must have data fields `private` and `public` which contain the gpg private and public key respectively.
+`publicKeys` is a list of public keys for the AttestationAuthority. Key rotation is supported by listing multiple keys.
+An image is attested if it has an attestation verifiable by ANY of the public keys.
 
 
-`publicKeys` is a list of public keys for the Kubernetes secret. Key rotation is supported by listing multiple keys.
-An image is attested if it has an attestation verifiable by ANY of the public keys.
+Each public key contains an ID. Signatures verified by BinAuthz must include the ID of the public key that can be used to
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply docs/resources.md]: Done.

File: pkg/kritis/review/validating_transport.go [hunk 28]
```diff
@@ -51,12 +53,17 @@ func (avt *AttestorValidatingTransport) validatePublicKey(pubKey v1beta1.PublicK
 	return nil
 }
 
-func validatePublicKeyType(pubKey v1beta1.PublicKey) error {
+// validatePublicKeyFields ensures that the appropriate fields of a PublicKey
+// are set given its KeyType.
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: Done.

File: pkg/kritis/review/validating_transport.go [hunk 29]
```diff
@@ -41,8 +41,10 @@ type AttestorValidatingTransport struct {
 	Attestor v1beta1.AttestationAuthority
 }
 
+// validatePublicKey makes sure that a PublicKey is specified correctly given
```
[the remark this answers is not in the document]
Comment (acamadeo) [reply pkg/kritis/review/validating_transport.go]: Done.
